Rewrite Comment Editor with React, install Webpack and React

noi5e opened the pull request:
Fixes #9175. Draft PR! 

---
(This issue is part of the larger Comment Editor Overhaul Project with Outreachy. Refer to Planning Issue #9069 for more context)

jywarren commented on app/controllers/notes_controller.rb:
```diff
@@ -44,6 +44,29 @@ def raw
   end
 
+  # takes an activerecord query, returns a plain array
+  def get_react_comments(comments_record, get_replies = false)
```
We may want to take a static version of the output from here and use it in a test; this could lay groundwork for externalizing the JS code in a library and then the JSON this generates would be the agreed-upon format for exchange between Rails and the JS lib; tested on both ends: here we'd test that Rails correctly generates the right format, while on the JS end, we'd test that the given format is readable and usable by the code. 

We can also annotate the JSON properties required in the README of the JS library, to explain what the parent app is supposed to supply! 

But we can wait a bit until this is more stable before doing too much of this. 

Great work!

jywarren commented on app/javascript/components/Comment.js:
```diff
@@ -0,0 +1,140 @@
+      timeCreatedString
+    } = this.props.comment;
+
+    let authorSection = [];
```
This is really nice. We might also store some of these subsections (strings) in a templates subfolder. I'm not familiar with React conventions though, is that often done to make things more concise and readable?

noi5e replied:
Yes, the main thing this is doing is for readability. There's a brief-ish writeup [here](https://www.freecodecamp.org/news/the-basics-of-destructuring-props-in-react-a196696f5477/).

Basically I 'destructure' the props here so that I can just refer to `replies` instead of `this.props.comment.replies`, as well as all the other props. Just saves effort typing, makes things more readable, and forces me to be clear about what props belong to, and need to be present in, whichever component.

jywarren commented on app/javascript/packs/application.js:
```diff
@@ -0,0 +1,22 @@
+// const images = require.context('../images', true)
+// const imagePath = (name) => images(name, true)
+
+console.log('Hello World from Webpacker')
```
Wheee!!! Ok, so not sure if i'm jumping the gun here but I've been wondering about webpack/rails asset pipeline integration. Some possible hints on this to bear in mind:

https://duckduckgo.com/?q=webpack+asset+pipeline&atb=v121-6&ia=images

jywarren replied:
I'm wondering if this is related to the test error we're seeing. But, the error seems relatively simple; it's asking us to fill out the `/public/packs-test/manifest.json` file with `application.js` (and possibly more than that?):

```
Minitest::UnexpectedError:         ActionView::Template::Error: Webpacker can't find application.js in /home/runner/work/plots2/plots2/public/packs-test/manifest.json. Possible causes:
        1. You want to set webpacker.yml value of compile to true for your environment
           unless you are using the `webpack -w` or the webpack-dev-server.
        2. webpack has not yet re-run to reflect updates.
        3. You have misconfigured Webpacker's config/webpacker.yml file.
        4. Your webpack configuration is not creating a manifest.
        Your manifest contains:
        {
        }
```

Is there a link to docs/guidance on the `manifest.json` file that could help us with this?

noi5e replied:
@jywarren This link seems relevant: https://github.com/reactjs/react-rails/issues/997#issuecomment-506258423

I do have a `manifest.json` set up locally, but I haven't been committing it. I added `public/packs` (which includes `manifest.json`) to my local `.gitignore`. I'm not 100% confident on this, but I think `manifest.json` isn't something to be committed, it's supposed to be generated with the instance of the app. See this for reference: https://stackoverflow.com/questions/48416457/should-we-commit-manifest-json-file-to-git-repo-when-using-rails-webpacker

I think what's supposed to happen is `rails generate` React commands have to be run on each particular instance of the app?

jywarren replied:
I see, ok, thanks!

I'll try this out but i have to run soon, so one thing we might do is use this `rails g webpack:install; rails g webpack:install:react; rails g react:install` on all the lines I added to `tests.yml` and `gitpod.yml`.

jywarren replied:
OK i tried this but didn't remove the prior `      run: ./public/lib/.bin/webpack-dev-server` -- but i can remove that too if it causes problems.

jywarren commented on app/views/notes/show.html.erb:
```diff
@@ -117,8 +117,20 @@
+        <!-- some methods aren't available in the controller, so we have to add them to the props object here -->
+        <!-- get_comment_react_props defined in application_helper.rb -->
+        <% comments = get_comment_react_props(@react_props[:comments]) %>
+        <%= react_component("CommentsContainer", { 
```
OK here may be where you need to use `<%=raw %>` to avoid escaping markup like `<p>`.

jywarren commented on bin/webpack:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env ruby
```
oh wow you are getting right into it, huh? Awesome!!! 

This may need a little more documentation in the main README since it's some pretty deep config going on in a new workflow. 

Great work!

noi5e replied:
@jywarren Wish I could take credit for being a config ninja, but it's really just the `react-rails` package doing everything behind the scenes.

Take a look at the [docs](https://github.com/reactjs/react-rails#get-started-with-webpacker), most of these config files were generated with just a few `rails generate` type commands.

It's pretty easy to set up in local development, I'm sure it would be easy for you to set this branch up locally on your system. However, it remains to be seen what a production deployment of this would be like.

jywarren reviewed (commented):
Keep up the great work! Big project! 💪

jywarren commented:
OK got it running!

![image](https://user-images.githubusercontent.com/24359/108913541-1b94ae00-75f8-11eb-9a11-7caedb12ef22.png)

I did have to run this to get it to work; it might also need `webpacker-cli` but let's see how it does:

`./node_modules/.bin/webpack-dev-server`

jywarren commented:
That's odd, i now see: `Error: Cannot find module 'ms'`

jywarren commented:
I see that https://github.com/rails/webpacker#development notes that using `webpack-dev-server` is supposed to be an alternative/backup, but it seemed to get me past the initial `manifest.json` error. But, if we get stuck on this `ms` module missing, we could go back and try to see how it'll work under the original default compilation setup, rather than this `webpack-dev-server` version. Still, a missing module seems easier to resolve than the mysterious `manifest.json` issue, at least to me.

jywarren commented on bin/webpack-dev-server:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env ruby
```
Hmm, so my accessing this at `/public/lib/.bin/webpack-dev-server` seems wrong, or maybe it's correct and we don't also need it here, esp if this copy of it would go out of date if we upgraded webpack? Maybe better to rely on the one in our /public/lib/ folder?

However, i think we are seeing some of these errors because we have an atypical location for our yarn dependencies, in public/lib instead of node_modules. I see this error when i try running `webpack-dev-server`:

```
gitpod /workspace/plots2 $ bin/webpack-dev-server 
internal/modules/cjs/loader.js:968
  throw err;
  ^

Error: Cannot find module 'webpack-cli/bin/config-yargs'
Require stack:
- /workspace/plots2/node_modules/webpack-dev-server/bin/webpack-dev-server.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:965:15)
    at Function.Module._load (internal/modules/cjs/loader.js:841:27)
    at Module.require (internal/modules/cjs/loader.js:1025:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/workspace/plots2/node_modules/webpack-dev-server/bin/webpack-dev-server.js:65:1)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
    at Module.load (internal/modules/cjs/loader.js:985:32)
    at Function.Module._load (internal/modules/cjs/loader.js:878:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/workspace/plots2/node_modules/webpack-dev-server/bin/webpack-dev-server.js'
  ]
}
```

noi5e replied:
@jywarren Thanks for doing so much work trying to set this up. It's nice to see some progress, although frustrating we can't get it working yet.

My understanding is that `webpack-dev-server` is just supposed to be used in development, which I guess GitHub CI & GitPod don't count as. So it does seem like that is more of a workaround than an actual solution.

Locally, I have a public/lib folder, and it's filled with dependencies. Is that normal? Are all the yarn dependencies supposed to be consolidated in `node_modules`? I don't know! This is just how the setup happened after `rails generate`.

I'm going to run `rails test functional` to see what happens.

There are also alternatives to the `react-rails` Gem that could work. Even after that swap, all the React components I write can probably just plug into that as written.

noi5e replied:
<img width="613" alt="Screen Shot 2021-02-23 at 2 49 08 PM" src="https://user-images.githubusercontent.com/4361605/108918529-60afe480-75e6-11eb-92ec-b930dc6c4080.png">

Still getting this when I open GitPod on this PR

noi5e replied:
Functional tests seem to be running fine locally BTW, I don't get the `manifest.json` errors that appear in CI

jywarren replied:
> in development, which I guess GitHub CI & GitPod don't count as. So it does seem like that is more of a workaround than an actual solution.

GitPod should be just like local development, so development mode, but you have a very good point on GitHub CI -- it's in test mode! We should search webpacker docs for how to get it running in test mode...

> Locally, I have a public/lib folder, and it's filled with dependencies. Is that normal? Are all the yarn dependencies supposed to be consolidated in node_modules? 

Yes, that's normal for plots2 - we were probably ahead of the curve on switching to Yarn so we have a custom location. We could consider switching to a more conventional location but only if it's essential, as that's a pretty big change just to get React running. I mean, we will probably have to do this for Rails 6 at some point in any case...

No worries here, this all seems like a pretty normal amount of difficulty for integrating a system like React+Webpack! Keep up the great work and i'm sure we'll crack it!

jywarren commented:
OK, so i believe adding `rails g webpack:install; rails g webpack:install:react; rails g react:install` as a setup command was critical as you suggested. Let's try adding that to gitpod.yml and to tests.yml where I had previously tried adding other commands. Great work on this!!!

noi5e commented:
@jywarren I tried to substitute the `webpack-dev-server` commands for `rails g` commands, but when I pushed the commits, I got turned down. I think maybe I don't have the authorization to change GitHub workflows?

<img width="1287" alt="Screen Shot 2021-02-23 at 10 00 14 PM" src="https://user-images.githubusercontent.com/4361605/108955250-b3a78d00-7622-11eb-9eba-14bdb1730ca4.png">

noi5e commented:
![Untitled 6](https://user-images.githubusercontent.com/4361605/109118619-6990db00-76f8-11eb-8edc-3c64a873e02e.gif)

I'm able to post comments now, and the comments appear on the page! I think this works for reply comments too, going to test it now.

Did some stuff on the controller end to make this happen, namely it rendering JSON. Also, I consolidated all the code in the `notes#show` view and `note_controller.rb` that was preparing the initial `comments` prop for React. 

Now, all the initial props generation is happening in the controller. It used to be happening in the controller AND the view. It was because I thought that some methods weren't accessible in the controller, but I've fixed it (still a lot I don't know about Rails).

jywarren commented on .github/workflows/tests.yml:
```diff
@@ -41,6 +41,10 @@ jobs:
     - name: Install JavaScript dependencies with Yarn
       run: yarn check || yarn install --frozen-lockfile;
+    - name: Run webpacker setup
+      run: rails g webpack:install && rails g webpack:install:react && rails g react:install
```
Hmm, @noi5e this caused:

```
/home/runner/work/plots2/plots2/vendor/bundle/ruby/2.6.0/gems/mysql2-0.5.3/lib/mysql2/client.rb:90:in `connect': Access denied for user 'runner'@'localhost' (using password: NO) (Mysql2::Error::ConnectionError)
```

That's a bit odd, what's going on there?

noi5e replied:
Hmm, that error does seem really weird. I'm not sure, it seems unrelated to React and webpacker? Going to run the tests again and see if it persists.

Here's the very start of that stacktrace:

```
from /home/runner/work/plots2/plots2/vendor/bundle/ruby/2.6.0/gems/railties-5.2.3/lib/rails/commands.rb:18:in `<top (required)>'
	from bin/rails:4:in `require'
	from bin/rails:4:in `<main>'
```

Doesn't it seem like something went wrong at a very fundamental level with `rails` and `railsties`? Could be a CI problem.

noi5e replied:
The tests are running again now, this appears in the stacktrace, so it seems like it's going to end similarly as the first run:
<img width="1090" alt="Screen Shot 2021-03-02 at 4 47 23 PM" src="https://user-images.githubusercontent.com/4361605/109735160-265ebe00-7b77-11eb-9550-2a537fc6ed9b.png">

However, looking further "upstream" it looks like the problem is caused by yarn not installing dependencies correctly? I wonder why?
<img width="1099" alt="Screen Shot 2021-03-02 at 4 46 49 PM" src="https://user-images.githubusercontent.com/4361605/109735134-1c3cbf80-7b77-11eb-82a6-a498bd35decb.png">

jywarren replied:
Hmm, yarn shouldn't relate to mysql though... i wonder. Let's see; certainly earlier commands related to the database ought to be using the correct mysql info, so maybe we adapt to how those worked...

jywarren commented on .github/workflows/tests.yml:
```diff
@@ -41,6 +41,11 @@ jobs:
     - name: Install JavaScript dependencies with Yarn
       run: yarn check || yarn install --frozen-lockfile;
+    - name: Run webpacker setup
+      env:
```
I added mysql info here. that should help!

jywarren replied:
Huh. Still didn't work, odd:

```
/home/runner/work/plots2/plots2/vendor/bundle/ruby/2.6.0/gems/mysql2-0.5.3/lib/mysql2/client.rb:90:in `connect': Access denied for user 'runner'@'localhost' (using password: NO) (Mysql2::Error::ConnectionError)
```

noi5e replied:
I'm going to help with this when I get a moment today...

Browsing the docs turns this up, which seems to be helpful: https://reactjs.org/blog/2013/07/30/use-react-and-jsx-in-ruby-on-rails.html

There's config I could change in `config/environments/development.rb` or perhaps `manifest.json` is committed? I'll have to look deeper.

noi5e replied:
Also this too, I bet config changes need to be made in `config/webpack`:https://stackoverflow.com/a/48769268/10977409

You said that GitHub CI is a `test` environment right?

jywarren replied:
I think i got it. tests are actually running now and failing due to assertions, not setup!

jywarren replied:
yay!!! functional and integration tests passed. i moved the fix to the system tests too so they should be good now as well! 🎉 🎉 🎉

noi5e replied:
@jywarren Nice! Whatever you did, did the trick!!! 🎉 🎈 🍰  I had to restart the CI a few times but currently all tests are passing:

<img width="574" alt="Screen Shot 2021-03-03 at 9 37 20 PM" src="https://user-images.githubusercontent.com/4361605/109916638-d2cc9d00-7c68-11eb-89bb-2ef86279b8ef.png">

And my code is live and working on GitPod!

<img width="697" alt="Screen Shot 2021-03-03 at 9 34 43 PM" src="https://user-images.githubusercontent.com/4361605/109916827-2343fa80-7c69-11eb-864e-9e4580edbd8e.png">

There is an issue with codeclimate:

<img width="830" alt="Screen Shot 2021-03-03 at 9 27 10 PM" src="https://user-images.githubusercontent.com/4361605/109916698-ef68d500-7c68-11eb-9e1b-4be4c81e3802.png">

It mentions `"Can't read property 'recommended' of undefined"`, which I think is caused by these changes I made in `.eslintrc` to install linters for React and React Hooks:

<img width="964" alt="Screen Shot 2021-03-03 at 9 26 19 PM" src="https://user-images.githubusercontent.com/4361605/109916773-0f989400-7c69-11eb-9063-90141032ca45.png">

I think I'm just going to push a commit to `.eslintrc` and see if that gets codeclimate working again, then... maybe this is mergeable?

noi5e replied:
Yeah that does work. codeclimate's displaying errors like normal now after my latest commits. I'll take a look at the errors tomorrow and see what I can address in this PR.

Weird that my changes to `.eslintrc` make ESLint work locally in VSCode, but they break codeclimate. Maybe I can revisit that in another PR.

jywarren replied:
OK that sounds good, thank you! If you can make an issue that'd be great, even a brief one.

noi5e replied:
Will make an issue for getting `.eslintrc` up and running again! I just pushed a rebase for this PR dropping [this commit](https://github.com/publiclab/plots2/commit/2456680208f5b45135d1b5785d8988ee88345e03) completely&mdash;the commit where I installed linters for React and React Hooks. I did this to undo the install to `package.json`.

jywarren commented:
OK, we are looking good here, then! I'd like to think through a few things before doing a merge:

1. documentation - because this introduces so mi'd like to ensure we have something in our Readme and maybe /docs/ folder discussing these changes, and perhaps linking to this PR to show all the files we touched. 
2. performance testing - do you have a good sense now of how this performs? either objectively on its own or in comparison to the previous comment code? 
3. this can help with performance -- can you try pushing this to the `unstable` branch and pinging @icarito to get it building at https://unstable.publiclab.org to test on a real-world database and production environment. I think it's not building right now but try pushing to the `unstable` branch and see if it triggers a build on https://jenkins.laboratoriopublico.org/

Thanks, @noi5e - this is a slightly different process due to the size and complexity of the PR as well as the structural changes -- i'm thinking of how we would provide support or guidance to new people looking to make changes to the commenting system. It can include a lot of links to React docs, we don't have to reinvent the wheel, but we should flag it to people that this is pretty different from how the rest of the codebase works.

Thanks!! Love to hear your thoughts on this!

jywarren commented on app/controllers/notes_controller.rb:
```diff
@@ -61,7 +61,41 @@ def show
+          :status => current_user[:status]
+        }
+
+        @react_props = {
```
Would this be better off inside the Comment model? Just an idea, maybe not though.

noi5e replied:
> Would this be better off inside the Comment model? Just an idea, maybe not though.

Maybe?? This is one of those things I'm not familiar with concerning Rails. From a performance standpoint, a method in the `Comment` model to `get_initial_react_props` would be faster than what I'm doing in the controller right now&mdash;right? I can add it to the list of optimizations.

jywarren commented:
One idea is -- what if we put a README.md inside the `app/javascript/components` folder - that would actually appear above the files in GitHub view, and could explain those files briefly?

Another thought is to rename this so it also indicates it's adding Webpack and React, for the record. Will do that now!

noi5e commented:
> Thanks, @noi5e - this is a slightly different process due to the size and complexity of the PR as well as the structural changes -- i'm thinking of how we would provide support or guidance to new people looking to make changes to the commenting system. It can include a lot of links to React docs, we don't have to reinvent the wheel, but we should flag it to people that this is pretty different from how the rest of the codebase works.

I hear you, going to work on this today!

noi5e commented:
I made commits last week that whittled down the codeclimate issues from 59 to 32. (Those 27 issues had to do with snake_casing and using Ruby 1.9 Hash syntax).

I am asking for a pass on the remaining 32 issues. I know that's a lot, but I can make a decent case. Here's a summary of the remaining codeclimate issues:

### `Issues with React Components`

<img width="533" alt="Screen Shot 2021-03-04 at 2 14 22 PM" src="https://user-images.githubusercontent.com/4361605/111039927-96760b00-83e5-11eb-954b-8b74ef67f024.png">

- **React components are too cognitively complex**
- **Function lengths exceed 25 lines**
- **Avoid deeply nested control flow statements**

Everything's still in flux, and I plan on doing a lot more restructuring of components. In particular, I'm going to restructure comments state so that replies are not nested within other comments. See https://github.com/publiclab/plots2/issues/9175#issuecomment-789455319. I think that will eliminate these issues.

### `babel.config.js files are too cognitively complex`

<img width="333" alt="Screen Shot 2021-03-04 at 2 14 36 PM" src="https://user-images.githubusercontent.com/4361605/111040661-dee2f800-83e8-11eb-9f3b-cc5773ec3061.png">

🤷 I don't think I can do anything about this one!

### `Issues with Comment Controllers`

<img width="326" alt="Screen Shot 2021-03-04 at 2 14 55 PM" src="https://user-images.githubusercontent.com/4361605/111040654-d68abd00-83e8-11eb-92b9-a0620834630e.png">

The solution for these kinds of issues will be to create entirely new routes like `/comment/REACT/create/46` instead of relying on the old route at `/comment/create/46`. That way, we can avoid weird conditional nesting like the following:

https://github.com/publiclab/plots2/blob/2c108c041e81dbc401f130c306f4604258a1e49b/app/controllers/comment_controller.rb#L118-L155

### `Issues with import, export, and require`

ESLint seems to have a problem parsing ES6 syntax. Again, not sure I can do anything about these:

<img width="361" alt="Screen Shot 2021-03-04 at 2 15 08 PM" src="https://user-images.githubusercontent.com/4361605/111040638-c1ae2980-83e8-11eb-8dc2-98d26911a3a6.png">

noi5e commented:
Pushing to `unstable` failed. This is the very end of [the stacktrace](https://jenkins.laboratoriopublico.org/job/Plots-Unstable/lastBuild/console): 

<img width="505" alt="Screen Shot 2021-03-13 at 11 09 25 AM" src="https://user-images.githubusercontent.com/4361605/111041353-a34a2d00-83ec-11eb-8565-9a27416ef4a6.png">

Tried scanning the rest of it. I'm not sure if this is the legitimate issue? I think maybe there might be some config issues to deal with, like the ones you did to get GitPod up and running.

noi5e commented:
Okay, I created a `REACT_COMMENTS.md` file in `/app/javascript/components`. See [here](https://github.com/publiclab/plots2/blob/7cd6b45a82e8ff710f92f4bc76cab7698be9ed88/app/javascript/components/REACT_COMMENTS.md).

That takes care of #1:

> 1. documentation - because this introduces so mi'd like to ensure we have something in our Readme and maybe /docs/ folder discussing these changes, and perhaps linking to this PR to show all the files we touched.

Regarding #2:

> 2. performance testing - do you have a good sense now of how this performs? either objectively on its own or in comparison to the previous comment code?

I think system tests are the way to go here (in another PR 😄)! Anecdotally speaking, React comments seem to perform just as well as the standard ones in all the regular ways. It seems fast. The only question I have is how the initial pageload of the research note is impacted while the server is bundling the initial JSON object, or doing things with webpacker. Even then, it still seems pretty fast.

> 3. this can help with performance -- can you try pushing this to the unstable branch and pinging @icarito to get it building at https://unstable.publiclab.org to test on a real-world database and production environment. I think it's not building right now but try pushing to the unstable branch and see if it triggers a build on https://jenkins.laboratoriopublico.org/

I need to do some more digging into the failed stacktrace, the GitPod configuration we did here. Will try to do a write-up in the next few days. Pointers definitely welcome though!!!

jywarren commented:
We can definitely bypass many of the codeclimate issues, no problem, esp. if they are for standard files or files whose lengthy you aren't solely responsible for. Not a big issue for me, thanks though!

For the unstable server it sounds like a common error we need to ask icarito to get unstuck. Can you ping him in the chatroom with the error message?

Finally i like the idea on performance testing. Sounds good! I also think that running locally you can see the breakdown of load speed for db vs. logic in the development logs... could be interesting to look at as well, for a quick assessment of the backend on initial page load. 

Thanks, @noi5e !!

noi5e commented:
Hey @icarito! Pinging you for help in getting this PR ready to merge. I'd like to demonstrate that it can work on `unstable`. Can you help us make the changes we need to get this working on `unstable`, and in production?

From a config standpoint, the main thing that this PR is doing is installing `rails-react` and `webpacker`. [Docs for the react-rails Gem](https://github.com/reactjs/react-rails)

I pushed this PR's branch to unstable, here's the [stacktrace for the failed build](https://jenkins.laboratoriopublico.org/job/Plots-Unstable/lastBuild/console).

The error messages in the stacktrace look very similar to config problems that we ran into setting up GitPod here. Jeffrey was able to get this running in GitPod by adding an `ENV` variable in [this commit](https://github.com/publiclab/plots2/pull/9176/commits/192489ddb6642443e188d0d6750679f2911a51fa). I think something similar on `unstable` and in production will probably work.

Also, if you look at the docs for `react-rails`, it seems important that a command to `rails g webpack:install && rails g webpack:install:react && rails g react:install` runs (just once). Jeffrey got this running in GitPod by doing that in [this commit](https://github.com/publiclab/plots2/pull/9176/commits/5d365749b51a1607d48688d6f17182c397477ad4). Those commands run every time a GitPod instance is created.

[This link from the rails-react docs](https://github.com/reactjs/react-rails#reactjs-versions) might also be pretty useful.

This kind of thing isn't my strong suit, but definitely happy to help do the research, make tweaks, or do whatever to get this up and running!

icarito commented:
I think maybe the key is the last message:
```

icarito commented:
[This](https://github.com/publiclab/plots2/blob/main/Dockerfile#L15) is the location where node is installed on our docker container. You would need to tweak that and confirm all else works for the next stable release of node you want to use.

noi5e commented:
@icarito Thanks! Yeah, I did see that error message earlier. Will try a commit to the Dockerfile.

noi5e commented:
@icarito Okay, I bumped node to v12 in the `Dockerfile`, that seems to take care of the webpacker issue.

The [build](https://jenkins.laboratoriopublico.org/job/Plots-Unstable/lastBuild/console) still fails. 

I think things start to get a little funny when yarn installs packages here:
```
info fsevents@2.3.2: The platform "linux" is incompatible with this module.
info "fsevents@2.3.2" is an optional dependency and failed compatibility check. Excluding it from installation.
```

I remember we encountered this same issue [here in this comment](https://github.com/publiclab/plots2/pull/9176#discussion_r586020933). I think this is the point where the fixes I mentioned [here](https://github.com/publiclab/plots2/pull/9176#issuecomment-800548034) become relevant.

noi5e commented:
@icarito Can you take a look at this again? I updated Node's version in `Dockerfile`, and it seems to take care of the error from earlier.

I'm thinking the next step is to run the `rails g webpack:install && rails g webpack:install:react && rails g react:install` on the `unstable` branch... Maybe in the `Dockerfile` again?

I just added this line and pushed to `unstable` again, let's see what happens: `RUN rails g webpack:install && rails g webpack:install:react && rails g react:install`

noi5e commented:
Looks like I got some working builds going on `unstable`! Everything seems to be working fine. 

See [the console output from the build](). There is a weird error `Could not find generator 'webpack:install:react'. Maybe you meant 'react:install', 'teaspoon:install' or 'atwho:install'` that doesn't seem to affect the success of the build at all. I think it has something to do with the order that the commands are run in.

I've tried running the commands in `start.sh` and `Makefile`. Maybe they have to run after a particular line to be just right?

noi5e commented:
From chat: 

> Okay great, I have a pretty good build going on unstable now. I figured out what was causing that error. I checked the react-rails documentation, and we were using 2 wrong commands in gitpod.yml and tests.yml, which was what I was copying-and-pasting.

> It's meant to be rails webpacker:install:react instead of rails g webpack:install:react. 🤦 @jywarren You might want to take note, because we currently have 2 out of 3 wrong commands in the GitPod & test flows!

> Only the third command rails generate react:install we had correct. What's weird is that seems to be the only one that's needed for a successful deploy? Anyway, this seems like it's working fine now.

I think this PR is pretty close to merge-ready? Here's the criteria again:

> 1. documentation - because this introduces so mi'd like to ensure we have something in our Readme and maybe /docs/ folder discussing these changes, and perhaps linking to this PR to show all the files we touched.
> 2. performance testing - do you have a good sense now of how this performs? either objectively on its own or in comparison to the previous comment code?
> 3. this can help with performance -- can you try pushing this to the unstable branch and pinging @icarito to get it building at https://unstable.publiclab.org to test on a real-world database and production environment. I think it's not building right now but try pushing to the unstable branch and see if it triggers a build on https://jenkins.laboratoriopublico.org/

1 is taken care of. 

2 - I'm going to take care of this very soon in the next couple of PRs. I'm going to select the most basic CRUD tests for comments, then run them on both Rails commenting and React commenting... Then build that into the system tests.

3 looks pretty good with the latest `unstable build`, but [here's the console output](https://jenkins.laboratoriopublico.org/job/Plots-Unstable/1098/console), in case we want to double-check it before merging.

jywarren commented:
Regarding performance system tests, if we feel they add too much time to the test runs, we can do the comparison in a PR, then based on the results, comment out the tests again before merging. Thank you!!!

jywarren commented:
🥁 🥁 🥁 🥁 🥁 🥁 🥁 🥁 🥁 🥁

jywarren commented:
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 

congratulations!!

RuthNjeri commented:
This is really amazing @noi5e, I would be interested in learning more about your work on this in the future....maybe we should have another 'open' call. Congrats 🚀

noi5e commented:
> This is really amazing @noi5e, I would be interested in learning more about your work on this in the future....maybe we should have another 'open' call. Congrats 🚀

@RuthNjeri, definitely, happy to set up a time to chat about it, just let me know!
